File: VirtualTap/src/tun_device.c
// TUN device creation and I/O for VirtualTap
// Platform-specific implementation for macOS, Linux, iOS

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <fcntl.h>
#include <errno.h>

#ifdef __APPLE__
#include <sys/socket.h>
#include <sys/ioctl.h>
#include <sys/kern_control.h>
#include <sys/sys_domain.h>
#include <net/if_utun.h>

// macOS utun device creation
int tun_device_create_macos(char *device_name, size_t device_name_size) {
    struct sockaddr_ctl addr;
    struct ctl_info info;
    int fd = -1;
    int unit_number;

    // Get utun control ID
    int temp_fd = socket(PF_SYSTEM, SOCK_DGRAM, SYSPROTO_CONTROL);
    if (temp_fd < 0) {
        fprintf(stderr, "[TUN] Failed to create socket: %s\n", strerror(errno));
        return -1;
    }

    memset(&info, 0, sizeof(info));
    strncpy(info.ctl_name, UTUN_CONTROL_NAME, sizeof(info.ctl_name));

    if (ioctl(temp_fd, CTLIOCGINFO, &info) < 0) {
        fprintf(stderr, "[TUN] CTLIOCGINFO failed: %s\n", strerror(errno));
        close(temp_fd);
        return -1;
    }
    close(temp_fd);

    // Try utun0-15
    for (unit_number = 0; unit_number < 16; unit_number++) {
        fd = socket(PF_SYSTEM, SOCK_DGRAM, SYSPROTO_CONTROL);
        if (fd < 0) {
            continue;
        }

        memset(&addr, 0, sizeof(addr));
        addr.sc_len = sizeof(addr);
        addr.sc_family = AF_SYSTEM;
        addr.ss_sysaddr = AF_SYS_CONTROL;
        addr.sc_id = info.ctl_id;
        addr.sc_unit = unit_number + 1; // utun0 = 1, utun1 = 2, etc.

        if (connect(fd, (struct sockaddr *)&addr, sizeof(addr)) < 0) {
            close(fd);
            fd = -1;
            continue;
        }

        // Success!
        break;
    }

    if (fd < 0) {
        fprintf(stderr, "[TUN] Failed to find available utun device\n");
        return -1;
    }

    // Get device name
    socklen_t optlen = (socklen_t)device_name_size;
    if (getsockopt(fd, SYSPROTO_CONTROL, UTUN_OPT_IFNAME, device_name, &optlen) < 0) {
        fprintf(stderr, "[TUN] UTUN_OPT_IFNAME failed: %s\n", strerror(errno));
        snprintf(device_name, device_name_size, "utun%d", unit_number);
    }

    // Set non-blocking
    int flags = fcntl(fd, F_GETFL, 0);
    if (flags >= 0) {
        fcntl(fd, F_SETFL, flags | O_NONBLOCK);
    }

    fprintf(stderr, "[TUN] Created TUN device: %s (fd=%d)\n", device_name, fd);
    return fd;
}

#elif defined(__linux__)
#include <sys/ioctl.h>
#include <linux/if.h>
#include <linux/if_tun.h>

// Linux TUN device creation
int tun_device_create_linux(char *device_name, size_t device_name_size) {
    struct ifreq ifr;
    int fd, err;

    fd = open("/dev/net/tun", O_RDWR);
    if (fd < 0) {
        fprintf(stderr, "[TUN] Failed to open /dev/net/tun: %s\n", strerror(errno));
        return -1;
    }

    memset(&ifr, 0, sizeof(ifr));
    ifr.ifr_flags = IFF_TUN | IFF_NO_PI; // TUN device, no protocol info

    if (device_name && device_name[0] != '\0') {
        strncpy(ifr.ifr_name, device_name, IFNAMSIZ);
    }

    err = ioctl(fd, TUNSETIFF, (void *)&ifr);
    if (err < 0) {
        fprintf(stderr, "[TUN] ioctl TUNSETIFF failed: %s\n", strerror(errno));
        close(fd);
        return -1;
    }

    // Copy device name back
    strncpy(device_name, ifr.ifr_name, device_name_size);
    
    // Set non-blocking
    int flags = fcntl(fd, F_GETFL, 0);
    if (flags >= 0) {
        fcntl(fd, F_SETFL, flags | O_NONBLOCK);
    }

    fprintf(stderr, "[TUN] Created TUN device: %s (fd=%d)\n", device_name, fd);
    return fd;
}
#endif

// Platform-independent TUN device creation
int tun_device_create(char *device_name, size_t device_name_size) {
#ifdef __APPLE__
    return tun_device_create_macos(device_name, device_name_size);
#elif defined(__linux__)
    return tun_device_create_linux(device_name, device_name_size);
#else
    fprintf(stderr, "[TUN] Unsupported platform\n");
    return -1;
#endif
}

// Close TUN device
void tun_device_close(int fd) {
    if (fd >= 0) {
        close(fd);
    }
}

// Read from TUN device (returns number of bytes read, or negative on error)
// On macOS, first 4 bytes are protocol info (0x00000002 = IPv4, 0x0000001e = IPv6)
int32_t tun_device_read(int fd, uint8_t *buffer, uint32_t buffer_size) {
    if (fd < 0 || !buffer) {
        return -1;
    }

    ssize_t n = read(fd, buffer, buffer_size);
    if (n < 0) {
        if (errno == EAGAIN || errno == EWOULDBLOCK) {
            return 0; // No data available (non-blocking)
        }
        return -1; // Error
    }

#ifdef __APPLE__
    // macOS utun prepends 4-byte protocol header
    // Strip it before returning to caller
    if (n >= 4) {
        memmove(buffer, buffer + 4, n - 4);
        return (int32_t)(n - 4);
    } else {
        return 0; // Invalid packet
    }
#else
    return (int32_t)n;
#endif
}

// Write to TUN device (returns number of bytes written, or negative on error)
// On macOS, we need to prepend 4-byte protocol info
int32_t tun_device_write(int fd, const uint8_t *packet, uint32_t packet_size) {
    if (fd < 0 || !packet || packet_size == 0) {
        return -1;
    }

#ifdef __APPLE__
    // macOS utun requires 4-byte protocol header
    // Determine protocol from IP version field
    uint32_t protocol = 0x00000002; // IPv4 (default)
    if (packet_size > 0) {
        uint8_t version = (packet[0] >> 4) & 0x0F;
        if (version == 6) {
            protocol = 0x0000001e; // IPv6
        }
    }

    // Build packet with protocol header
    uint8_t buf[2048];
    if (packet_size + 4 > sizeof(buf)) {
        return -1; // Packet too large
    }

    // Write protocol in network byte order
    buf[0] = 0;
    buf[1] = 0;
    buf[2] = 0;
    buf[3] = (protocol == 0x0000001e) ? 0x1e : 0x02;
    
    memcpy(buf + 4, packet, packet_size);

    ssize_t n = write(fd, buf, packet_size + 4);
    if (n < 0) {
        if (errno == EAGAIN || errno == EWOULDBLOCK) {
            return 0; // Would block
        }
        return -1; // Error
    }

    // Return bytes written (excluding protocol header)
    return (int32_t)(n > 4 ? n - 4 : 0);
#else
    ssize_t n = write(fd, packet, packet_size);
    if (n < 0) {
        if (errno == EAGAIN || errno == EWOULDBLOCK) {
            return 0;
        }
        return -1;
    }
    return (int32_t)n;
#endif
}

File: libs/virtualtap/src/tun_device.rs
//! Pure Rust TUN device implementation for macOS
//! 
//! Creates utun devices using ioctl syscalls without C dependencies

use std::os::unix::io::RawFd;
use std::ffi::CString;

const UTUN_CONTROL_NAME: &str = "com.apple.net.utun_control";
const MAX_PACKET_SIZE: usize = 2048;

#[repr(C)]
struct CtlInfo {
    ctl_id: u32,
    ctl_name: [u8; 96],
}

#[repr(C)]
struct SockaddrCtl {
    sc_len: u8,
    sc_family: u8,
    ss_sysaddr: u16,
    sc_id: u32,
    sc_unit: u32,
    sc_reserved: [u32; 5],
}

pub struct TunDevice {
    fd: RawFd,
    name: String,
}

impl TunDevice {
    /// Create a new TUN device (utun) on macOS
    pub fn create() -> Result<Self, String> {
        unsafe {
            // Create socket for utun
            let fd = libc::socket(libc::PF_SYSTEM, libc::SOCK_DGRAM, libc::SYSPROTO_CONTROL);
            if fd < 0 {
                return Err(format!("Failed to create socket: {}", std::io::Error::last_os_error()));
            }

            // Get control ID for utun
            let mut ctl_info = std::mem::zeroed::<CtlInfo>();
            let control_name = CString::new(UTUN_CONTROL_NAME).unwrap();
            let name_bytes = control_name.as_bytes_with_nul();
            ctl_info.ctl_name[..name_bytes.len()].copy_from_slice(name_bytes);

            const CTLIOCGINFO: libc::c_ulong = 0xc0644e03; // ioctl code for CTLIOCGINFO
            
            if libc::ioctl(fd, CTLIOCGINFO, &mut ctl_info) < 0 {
                libc::close(fd);
                return Err(format!("Failed to get control info: {}", std::io::Error::last_os_error()));
            }

            // Connect to utun control (unit 0 = let kernel pick)
            let mut addr = std::mem::zeroed::<SockaddrCtl>();
            addr.sc_len = std::mem::size_of::<SockaddrCtl>() as u8;
            addr.sc_family = libc::AF_SYSTEM as u8;
            addr.ss_sysaddr = libc::AF_SYS_CONTROL as u16;
            addr.sc_id = ctl_info.ctl_id;
            addr.sc_unit = 0; // Let kernel pick unit number

            if libc::connect(
                fd,
                &addr as *const _ as *const libc::sockaddr,
                std::mem::size_of::<SockaddrCtl>() as u32,
            ) < 0 {
                libc::close(fd);
                return Err(format!("Failed to connect to utun control: {}", std::io::Error::last_os_error()));
            }

            // Get the assigned unit number
            let mut unit: u32 = 0;
            let mut unit_len: libc::socklen_t = std::mem::size_of::<u32>() as u32;
            
            const SYSPROTO_CONTROL: i32 = 2;
            const UTUN_OPT_IFNAME: i32 = 2;
            
            if libc::getsockopt(
                fd,
                SYSPROTO_CONTROL,
                UTUN_OPT_IFNAME,
                &mut unit as *mut _ as *mut libc::c_void,
                &mut unit_len,
            ) < 0 {
                // If we can't get the name, just use unit 0
                unit = 0;
            }

            let name = format!("utun{}", unit);
            
            // Set non-blocking mode for async I/O
            let flags = libc::fcntl(fd, libc::F_GETFL, 0);
            if flags >= 0 {
                let _ = libc::fcntl(fd, libc::F_SETFL, flags | libc::O_NONBLOCK);
            }
            
            Ok(TunDevice { fd, name })
        }
    }

    pub fn name(&self) -> &str {
        &self.name
    }

    pub fn fd(&self) -> RawFd {
        self.fd
    }

    /// Read an IP packet from the TUN device
    pub fn read(&self) -> Result<Option<Vec<u8>>, String> {
        let mut buffer = vec![0u8; MAX_PACKET_SIZE];
        
        unsafe {
            let n = libc::read(self.fd, buffer.as_mut_ptr() as *mut libc::c_void, buffer.len());
            
            if n < 0 {
                let err = std::io::Error::last_os_error();
                if err.kind() == std::io::ErrorKind::WouldBlock {
                    return Ok(None); // Non-blocking, no data available
                }
                return Err(format!("TUN read error: {}", err));
            }
            
            if n == 0 {
                return Ok(None);
            }

            // macOS utun prepends 4-byte protocol family header
            if n < 4 {
                return Err("Packet too short".to_string());
            }

            // Skip the 4-byte header and return the IP packet
            buffer.truncate(n as usize);
            Ok(Some(buffer[4..].to_vec()))
        }
    }

    /// Write an IP packet to the TUN device
    pub fn write(&self, packet: &[u8]) -> Result<usize, String> {
        // macOS utun requires 4-byte protocol family header
        let mut buffer = Vec::with_capacity(4 + packet.len());
        
        // Determine protocol family from IP version
        let family = if !packet.is_empty() {
            let version = (packet[0] >> 4) & 0x0F;
            match version {
                4 => libc::AF_INET as u32,
                6 => libc::AF_INET6 as u32,
                _ => return Err("Invalid IP version".to_string()),
            }
        } else {
            return Err("Empty packet".to_string());
        };

        // Prepend 4-byte family header (network byte order)
        buffer.extend_from_slice(&family.to_be_bytes());
        buffer.extend_from_slice(packet);

        unsafe {
            let n = libc::write(self.fd, buffer.as_ptr() as *const libc::c_void, buffer.len());
            
            if n < 0 {
                return Err(format!("TUN write error: {}", std::io::Error::last_os_error()));
            }

            // Return number of IP packet bytes written (excluding 4-byte header)
            Ok((n as usize).saturating_sub(4))
        }
    }
}

impl Drop for TunDevice {
    fn drop(&mut self) {
        unsafe {
            libc::close(self.fd);
        }
    }
}

unsafe impl Send for TunDevice {}
unsafe impl Sync for TunDevice {}
